dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java

```diff
@@ -201,4 +200,35 @@ public void deleteByDsoAndTypeNotEqualsTo(Context context, DSpaceObject dso, Str
         query.setParameter("rptype", type);
         query.executeUpdate();
     }
+
+    @Override
+    public List<ResourcePolicy> findByDSoAndActionExceptRpType(Context context, DSpaceObject dso, int action,
+                                                               String rpType) throws SQLException {
+        CriteriaBuilder criteriaBuilder = getCriteriaBuilder(context);
+        CriteriaQuery criteriaQuery = getCriteriaQuery(criteriaBuilder, ResourcePolicy.class);
+
+        Root<ResourcePolicy> resourcePolicyRoot = criteriaQuery.from(ResourcePolicy.class);
+        criteriaQuery.select(resourcePolicyRoot);
+        if (rpType != null) {
+            criteriaQuery.where(
+                criteriaBuilder.and(criteriaBuilder.equal(resourcePolicyRoot.get(ResourcePolicy_.dSpaceObject), dso),
+                                    criteriaBuilder.equal(resourcePolicyRoot.get(ResourcePolicy_.actionId), action),
+                                    criteriaBuilder.or(
+                                            criteriaBuilder.notEqual(resourcePolicyRoot.get(ResourcePolicy_.rptype),
```

Comment:
Because you query the "rptype" column here with an "OR" query, I think this is the first method where "rptype" is queried "standalone". This means the database will probably need an index on this column to remain performant.

Can you please check if this index is already present and if not, create it via Flyway?

Reply:
yes, it is already in place 
https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/postgres/V6.0_2016.04.04__DS-3086-OAI-Performance-fix.sql#L22